### ISSUE TITLE
Multiple sample description boxes are not handled

### ISSUE THREAD

**Issue by SingingTree** [mozilla/mp4parse-rust#160]
When we parse the stsd box, we iterate over the sample descriptions contained in the stsd. Based on the [code](https://github.com/mozilla/mp4parse-rust/blob/master/mp4parse/src/lib.rs#L2072) for parsing at the moment, I suspect that most mp4s contain a single description.

However, while working on support for handling parsing for the cbcs encryption scheme I've noticed that the test media I've created using [shaka-packager](https://github.com/google/shaka-packager) has multiple description boxes.

Here are video files I created for testing. The init is the primary one of interest, but I've included the other least they're useful: [bipbop_cbcs_video_files.zip](https://github.com/mozilla/mp4parse-rust/files/2462614/bipbop_cbcs_video_files.zip)

The files were created (along with some audio files and a manifest) from mozilla-central's bipbop.mp4 via the following:
```
packager-win.exe in=bipbop.mp4,stream=audio,init_segment=bipbop_cbcs_audio_init.mp4,segment_template=bipbop_cbcs_audio_$Number$.m4s in=bipbop.mp4,stream=video,init_segment=bipbop_cbcs_video_init.mp4,segment_template=bipbop_cbcs_video_$Number$.m4s --protection_scheme cbcs --enable_raw_key_encryption  --keys label=:key_id=7e571d047e571d047e571d047e571d21:key=7e5744447e5744447e5744447e574421 --iv 11223344556677889900112233445566 --generate_static_mpd --mpd_output bipbop_cbcs.mpd
```

The first issue I'm hitting on is that the codec_type we set on the track is set based on the [last description parsed](https://github.com/mozilla/mp4parse-rust/blob/master/mp4parse/src/lib.rs#L2083). This means for the test media above, the video track is labelled as `H264`, rather than `EncryptedVideo`.

Furthermore, we [only store the first encountered description](https://github.com/mozilla/mp4parse-rust/blob/master/mp4parse/src/lib.rs#L2096) (contrast this that the type appears to be derived from the last encountered one).

I'm no expert on the format, and am still trying to understand all the ins and outs of how these descriptions are used, but I imagine if both descriptions are present we should store them, as they may be referenced by fragments later.

While I continue trying to get my head around the details, I figured I'd create an issue for discussion on how to handle this. @kinetiknz, @jyavenard, do you have thoughts?

**Comment by SingingTree** [comment]
After more of a look, I think I understand better how this data is used by fragments:

- The Track Fragment Header Box (`tfhd`) can use its `sample_description_index` + `track_ID` to reference different Sample Descriptions.
- For the above files it appears this is done to reference different information for the clear fragment and the encrypted fragment.
  - It so happens that most of the info is the same between those descriptions, but I see no reason why they couldn't deviate further.

Right now it looks like we [drop the descriptions returned from `read_stsd`](https://github.com/mozilla/mp4parse-rust/blob/f6032a118aa498525145adf611cd7b3bec0e0216/mp4parse/src/lib.rs#L962) and [keep the first one we encounter on the track struct](https://github.com/mozilla/mp4parse-rust/blob/f6032a118aa498525145adf611cd7b3bec0e0216/mp4parse/src/lib.rs#L2097). If I understand correctly, this means we can't do the `tfhd` lookup noted above.

I propose we rework `read_stsd` so it no longer takes a `track` arg, and instead assign the descriptions inside [`read_stbl`](https://github.com/mozilla/mp4parse-rust/blob/f6032a118aa498525145adf611cd7b3bec0e0216/mp4parse/src/lib.rs#L963) to the track.

This is likely to change the API, and make handling track data more complex. However, this would ultimately lead to the ability to handle a greater variety of files, which includes the immediate need in order to parse shaka packaged cbcs video.

Will start work on a PR unless there are concerns about the above.

**Comment by kinetiknz** [comment]
Closed via #163
